In [1]:
# Import dependencies
import pandas as pd
import numpy as np
import re
import time

In [2]:
# Create DataFrame from CSV file
nlp_df = pd.read_csv('../Data/nlp_df.csv')
nlp_df = nlp_df.drop(['non_alpha_words'], axis=1)
nlp_df.head(3)

song                 song_id  \
0                                 willow  0lx2cLdOt3piJbcaXIV74f   
1  Stay Next To Me (with Chelsea Cutler)  6SGG5AxHShqSYiV9fCWpZz   
2                            WITHOUT YOU  27OeeYzk6klgBh83TSvGMA   

          artist               artist_id category  category_id  popularity  \
0   Taylor Swift  06HL4z0CvFAxyc27GXpf02      pop            8          93   
1     Quinn XCII  3ApUX1o6oSz321MMECyIYd      pop            8          78   
2  The Kid LAROI  2tIP7SsRs7vjIcLrU85W8J      pop            8          95   

                           genres  audio_ft_danceability  audio_ft_energy  \
0                ['dance', 'pop']                  0.392            0.574   
1  ['indie', 'pop', 'electropop']                  0.581            0.584   
2                  ['australian']                  0.662            0.413   

   ...  audio_ft_acousticness  audio_ft_instrumentalness  audio_ft_liveness  \
0  ...                 0.8330                    0.00179              0.145   
1  ...                 0.0805                    0.00000              0.366   
2  ...                 0.2130                    0.00000              0.134   

   audio_ft_valence  audio_ft_tempo  audio_ft_duration_ms  \
0             0.529          81.112              214707.0   
1             0.756         179.954              206046.0   
2             0.467          93.005              161385.0   

   audio_ft_time_signature                                             lyrics  \
0                      4.0   Im like the water when your ship rolled in th...   
1                      4.0   Didnt even wanna go out whyd you call me ? Iv...   
2                      4.0   You cut out a piece of me and now I bleed int...   

                                            filtered   language  
0  ['', 'im', 'like', 'water', 'ship', 'rolled', ...  (1, 'en')  
1  ['', 'didnt', 'even', 'wanna', 'go', 'whyd', '...  (1, 'en')  
2  ['', 'cut', 'piece', 'bleed', 'internally', 'l...  (1, 'en')  

[3 rows x 24 columns]

In [3]:
# Create a list of all words, word counts, unique word counts, and filtered words
t0 = time.time()
words_list = []
word_counts = []
unique_word_counts = []
filtered_words_list = []
for index, row in nlp_df.iterrows():
    filtered_words = row['filtered']
    filtered_words = filtered_words.replace(',', '').replace("'", '')
    filtered_words = filtered_words.replace('[', '').replace(']', '')
    filtered_words = filtered_words.replace('#', '').replace('&nbsp', '')
    filtered_words = filtered_words.replace('?', '? ').replace(',', '')
    filtered_words = filtered_words.replace('/', ' ')
    filtered_words = filtered_words.replace('\\u200a', '').replace('\\u200b', '')
    filtered_words = filtered_words.replace('\\u2063', '').replace('\u202f', '')
    filtered_words = filtered_words.replace('\\u2028', ' ').replace('\\u2008', ' ')
    while ('\\u200e' in filtered_words) or ('\\xa0' in filtered_words):
        filtered_words = filtered_words.replace('\\u200e', '')
        filtered_words = filtered_words.replace('\\xa0', '')
    filtered_words_list.append(filtered_words)
    words = filtered_words.strip().split(' ')
    word_counts.append(len(words))
    unique_words = list(set(words))
    unique_word_counts.append(len(unique_words))
    words_list.extend(unique_words)
word_columns = list(set(words_list))
t1 = time.time()
print(f'Run time: {t1-t0} seconds')
len(word_columns)

Run time: 1.387387752532959 seconds


36234

In [4]:
# Add word_count and unique_word_count columns and replace the filtered column
nlp_df['word_count'] = word_counts
nlp_df['unique_word_count'] = unique_word_counts
nlp_df['filtered'] = filtered_words_list
nlp_df.head(1)

song                 song_id        artist               artist_id  \
0  willow  0lx2cLdOt3piJbcaXIV74f  Taylor Swift  06HL4z0CvFAxyc27GXpf02   

  category  category_id  popularity            genres  audio_ft_danceability  \
0      pop            8          93  ['dance', 'pop']                  0.392   

   audio_ft_energy  ...  audio_ft_liveness  audio_ft_valence  audio_ft_tempo  \
0            0.574  ...              0.145             0.529          81.112   

   audio_ft_duration_ms  audio_ft_time_signature  \
0              214707.0                      4.0   

                                              lyrics  \
0   Im like the water when your ship rolled in th...   

                                            filtered   language  word_count  \
0   im like water ship rolled night rough surface...  (1, 'en')         227   

   unique_word_count  
0                 87  

[1 rows x 26 columns]

In [5]:
# Remove songs with less than 25 unique words
nlp_df = nlp_df[nlp_df['unique_word_count']>=25]
len(nlp_df)

8041

# Examine word counts by genre

In [7]:
# Remove the language column
nlp_df = nlp_df.drop('language', axis=1)

In [6]:
# Drop songs with incorrect lyrics from the DataFrame
nlp_df = nlp_df[(nlp_df['unique_word_count']!=3878) & (nlp_df['unique_word_count']!=1153)]
nlp_df = nlp_df[(nlp_df['unique_word_count']!=1000) & (nlp_df['unique_word_count']!=880)]
nlp_df = nlp_df[(nlp_df['unique_word_count']!=842) & (nlp_df['unique_word_count']!=607)]
nlp_df = nlp_df[(nlp_df['unique_word_count']!=569) & (nlp_df['word_count']!=1383)]
len(nlp_df)

8031

In [8]:
nlp_df['popularity'].describe()

count    8031.000000
mean       47.177188
std        22.731742
min         0.000000
25%        37.000000
50%        51.000000
75%        63.000000
max        99.000000
Name: popularity, dtype: float64

In [9]:
# Print the distributions of word counts for each quartile
word_distributions = {}
quartile1_unique_words = nlp_df[nlp_df['popularity']<=37]['word_count'].describe()
word_distributions['first_quartile'] = quartile1_unique_words

quartile2_unique_words = nlp_df[(nlp_df['popularity']>37) & (nlp_df['popularity']<=51)]['word_count'].describe()
word_distributions['second_quartile'] = quartile2_unique_words

quartile3_unique_words = nlp_df[(nlp_df['popularity']>51) & (nlp_df['popularity']<63)]['word_count'].describe()
word_distributions['third_quartile'] = quartile3_unique_words

quartile4_unique_words = nlp_df[nlp_df['popularity']>=63]['word_count'].describe()
word_distributions['fourth_quartile'] = quartile4_unique_words
word_distributions

{'first_quartile': count    2067.000000
 mean      154.472182
 std        78.391665
 min        26.000000
 25%       102.000000
 50%       141.000000
 75%       189.000000
 max       694.000000
 Name: word_count, dtype: float64,
 'second_quartile': count    1949.000000
 mean      161.593125
 std        85.961241
 min        36.000000
 25%       105.000000
 50%       142.000000
 75%       193.000000
 max       823.000000
 Name: word_count, dtype: float64,
 'third_quartile': count    1870.000000
 mean      191.383422
 std       104.526280
 min        32.000000
 25%       125.000000
 50%       169.000000
 75%       223.000000
 max       832.000000
 Name: word_count, dtype: float64,
 'fourth_quartile': count    2145.000000
 mean      205.691375
 std        97.987791
 min        33.000000
 25%       143.000000
 50%       185.000000
 75%       239.000000
 max       909.000000
 Name: word_count, dtype: float64}

In [10]:
# Create a dictionary of mean word distributions
categories = list(word_distributions.keys())
means = []
for category in word_distributions:
    means.append(word_distributions[category]['mean'])
mean_unique_word_counts = dict(zip(categories, means))
mean_unique_word_counts = dict(sorted(mean_unique_word_counts.items(), key=lambda item: item[1], reverse=True))
mean_unique_word_counts

{'fourth_quartile': 205.6913752913753,
 'third_quartile': 191.38342245989304,
 'second_quartile': 161.59312467932273,
 'first_quartile': 154.47218190614416}

In [11]:
# Find songs with the highest and lowest word counts
print('Song with the most words:')
display(nlp_df[nlp_df['word_count']==max(nlp_df['word_count'])][['song', 'artist', 'category', 'popularity', 'lyrics', 'word_count']])

print('Song with the least words:')
display(nlp_df[nlp_df['word_count']==min(nlp_df['word_count'])][['song', 'artist', 'category', 'popularity', 'lyrics', 'word_count']])


Song with the most words:


song  artist category  popularity  \
1351  Rap God  Eminem   hiphop          64   

                                                 lyrics  word_count  
1351   Look I was gonna go easy on you not to hurt y...         909

Song with the least words:


song     artist category  popularity  \
2547  Golf Hill Drive  Boys Life     rock          25   

                                                 lyrics  word_count  
2547  Stuck in this life thats not my own And whoeve...          26

In [12]:
# Create DataFrame of word count distributions
ranges = []
for category in word_distributions:
    ranges.append(word_distributions[category]['max'] - word_distributions[category]['min'])

columns = ['count', 'mean', 'std', 'min', '25%', '50%', '75%', 'max', 'range']
df_rows = []
for index, category in enumerate(word_distributions):
    distribution = list(word_distributions[category])
    distribution.extend([ranges[index]])
    df_rows.append(distribution)
popularity_df = pd.DataFrame(df_rows, index=list(word_distributions.keys()), columns=columns)
popularity_df

count        mean         std   min    25%    50%    75%  \
first_quartile   2067.0  154.472182   78.391665  26.0  102.0  141.0  189.0   
second_quartile  1949.0  161.593125   85.961241  36.0  105.0  142.0  193.0   
third_quartile   1870.0  191.383422  104.526280  32.0  125.0  169.0  223.0   
fourth_quartile  2145.0  205.691375   97.987791  33.0  143.0  185.0  239.0   

                   max  range  
first_quartile   694.0  668.0  
second_quartile  823.0  787.0  
third_quartile   832.0  800.0  
fourth_quartile  909.0  876.0

In [13]:
# Save DataFrame to CSV
popularity_df.to_csv('../Data/Analysis/popularity_by_words.csv')

# Conclusions about word distributions by popularity
- The songs with the lowest 25% popularities have the lowest mean word count of 154 words
- The songs in the second quartile of popularity have a mean word count of 162 words
- The songs in the third quartile of popularity have a mean word count of 191 words
- The songs with the highest 25% popularities ahve the highest mean word count fo 206 words
- The song with the highest word count has a popularity of 64:
    - Song name: Rap God 
    - Artist: Eminem
    - Word count: 909 words
- The song with the lowest word count has a popularity of 25:
    - Song name: Golf Hill Drive 
    - Artist: Boys Life
    - Word count: 26 words

# Examine unique word counts by genre

In [14]:
# Print the distributions of unique word counts for each quartile
unique_word_distributions = {}
quartile1_unique_words = nlp_df[nlp_df['popularity']<=37]['unique_word_count'].describe()
unique_word_distributions['first_quartile'] = quartile1_unique_words

quartile2_unique_words = nlp_df[(nlp_df['popularity']>37) & (nlp_df['popularity']<=52)]['unique_word_count'].describe()
unique_word_distributions['second_quartile'] = quartile2_unique_words

quartile3_unique_words = nlp_df[(nlp_df['popularity']>52) & (nlp_df['popularity']<63)]['unique_word_count'].describe()
unique_word_distributions['third_quartile'] = quartile3_unique_words

quartile4_unique_words = nlp_df[nlp_df['popularity']>=63]['unique_word_count'].describe()
unique_word_distributions['fourth_quartile'] = quartile4_unique_words
unique_word_distributions

{'first_quartile': count    2067.000000
 mean       69.169328
 std        38.362064
 min        25.000000
 25%        47.000000
 50%        62.000000
 75%        79.000000
 max       486.000000
 Name: unique_word_count, dtype: float64,
 'second_quartile': count    2120.000000
 mean       73.972642
 std        44.358010
 min        25.000000
 25%        49.000000
 50%        63.000000
 75%        82.000000
 max       503.000000
 Name: unique_word_count, dtype: float64,
 'third_quartile': count    1699.000000
 mean       82.720424
 std        52.921538
 min        25.000000
 25%        52.500000
 50%        68.000000
 75%        92.000000
 max       544.000000
 Name: unique_word_count, dtype: float64,
 'fourth_quartile': count    2145.000000
 mean       82.702098
 std        47.988176
 min        25.000000
 25%        55.000000
 50%        70.000000
 75%        92.000000
 max       539.000000
 Name: unique_word_count, dtype: float64}

In [16]:
# Create a dictionary of mean unique word distributions
categories = list(unique_word_distributions.keys())
means = []
for category in unique_word_distributions:
    means.append(unique_word_distributions[category]['mean'])
mean_unique_word_counts = dict(zip(categories, means))
mean_unique_word_counts = dict(sorted(mean_unique_word_counts.items(), key=lambda item: item[1], reverse=True))
mean_unique_word_counts

{'third_quartile': 82.72042377869334,
 'fourth_quartile': 82.7020979020979,
 'second_quartile': 73.97264150943396,
 'first_quartile': 69.1693275278181}

In [17]:
# Find songs with the highest and lowest  word counts
print('Song with the most unique words:')
display(nlp_df[nlp_df['unique_word_count']==max(nlp_df['unique_word_count'])][['song', 'artist', 'category', 'popularity', 'lyrics', 'unique_word_count']])

print('Popularities of songs that have the minimum unique word count of 25 words:')
print(set(nlp_df[nlp_df['unique_word_count']==min(nlp_df['unique_word_count'])]['popularity']), '\n')

print(nlp_df[nlp_df['unique_word_count']==min(nlp_df['unique_word_count'])]['popularity'].describe())

min_word_count_df = nlp_df[nlp_df['unique_word_count']==min(nlp_df['unique_word_count'])]
print(f'{len(min_word_count_df)} songs have the minimum unique word count of 25 words.')


Song with the most unique words:


song artist category  \
1262  Momentum (feat. Black Thought & Benny The Butc...   Russ   hiphop   

      popularity                                             lyrics  \
1262          61   Yeah yeah   Dont compare me to employees who ...   

      unique_word_count  
1262                544

Popularities of songs that have the minimum unique word count of 25 words:
{0, 4, 25, 26, 30, 31, 35, 38, 42, 43, 45, 46, 48, 49, 60, 61, 62, 70, 84} 

count    25.000000
mean     36.160000
std      24.463715
min       0.000000
25%      25.000000
50%      42.000000
75%      49.000000
max      84.000000
Name: popularity, dtype: float64
25 songs have the minimum unique word count of 25 words.


In [18]:
# Create DataFrame of unique word count distributions
ranges = []
for category in unique_word_distributions:
    ranges.append(unique_word_distributions[category]['max'] - unique_word_distributions[category]['min'])

columns = ['count', 'mean', 'std', 'min', '25%', '50%', '75%', 'max', 'range']
df_rows = []
for index, category in enumerate(unique_word_distributions):
    distribution = list(unique_word_distributions[category])
    distribution.extend([ranges[index]])
    df_rows.append(distribution)
unique_popularity_df = pd.DataFrame(df_rows, index=list(unique_word_distributions.keys()), columns=columns)
unique_popularity_df

count       mean        std   min   25%   50%   75%    max  \
first_quartile   2067.0  69.169328  38.362064  25.0  47.0  62.0  79.0  486.0   
second_quartile  2120.0  73.972642  44.358010  25.0  49.0  63.0  82.0  503.0   
third_quartile   1699.0  82.720424  52.921538  25.0  52.5  68.0  92.0  544.0   
fourth_quartile  2145.0  82.702098  47.988176  25.0  55.0  70.0  92.0  539.0   

                 range  
first_quartile   461.0  
second_quartile  478.0  
third_quartile   519.0  
fourth_quartile  514.0

In [19]:
# Save DataFrame to CSV
unique_popularity_df.to_csv('../Data/Analysis/popularity_by_unique_words.csv')

# Conclusions about unique word distributions by popularity
- The songs with the lowest 25% popularities have the lowest mean unique word count of 69 words
- The song with the highest unique word count has a popularity of 61:
    - Song name: Momentum 
    - Artist: Russ
    - Unique word count: 544 words
- The average popularity of the 25 songs with minimum unique word count was 36
    - The highest popularity of the songs with the minimum unique word count was 84
    - The lowest popularity of the songs with the minimum unique word count was 0